### ISSUE TITLE
Version number in npm state name does not result in correct version being installed

### ISSUE THREAD

**Issue by hundt** [saltstack/salt#18647]
If you add a version number to the state name in an npm state (like coffee-script@1.0.1 from the docs) then `states/npm.py` will correctly check whether that version is installed, but it does not actually pass that version to npm when it tries to install. I believe the offending code is line 170 of `npm.py`:

`cmd_args['pkg'] = pkg_name  # note that version is not included`

**Comment by notpeter** [comment]
This was a regression introduced in 2014.7.
Version 2014.1.13 handles this correctly.

**Comment by rallytime** [comment]
Thanks for the extra info, @notpeter!

**Comment by rallytime** [comment]
I went to take a look at this today and I think this has been fixed as I cannot reproduce the error. Here are my findings:
```
# cat npm-test.sls
coffee-script@1.0.1:
  npm.installed: []

# salt nt-master state.sls npm-test
nt-master:
----------
          ID: coffee-script@1.0.1
    Function: npm.installed
      Result: True
     Comment: Package(s) 'coffee-script@1.0.1' successfully installed
     Started: 19:40:44.853536
    Duration: 2387.593 ms
     Changes:
              ----------
              new:
                  - coffee-script@1.0.1
              old:

Summary
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1

# salt nt-master npm.list
nt-master:
    ----------
    coffee-script:
        ----------
        from:
            coffee-script@1.0.1
        version:
            1.0.1
```
I then removed the coffee-script installation and edited my state file to not contain a version, just to make sure 1.0.1 wasn't the default install. Sure enough, it's installing a new version like so:
```
# salt nt-master state.sls npm-test
nt-master:
----------
          ID: coffee-script
    Function: npm.installed
      Result: True
     Comment: Package(s) 'coffee-script' successfully installed
     Started: 19:41:47.422252
    Duration: 1920.605 ms
     Changes:
              ----------
              new:
                  - coffee-script
              old:

Summary
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1

# salt nt-master npm.list
nt-master:
    ----------
    coffee-script:
        ----------
        from:
            coffee-script@
        version:
            1.9.1
```
This is fixed on the latest release, 2014.7.2:
```
# salt --versions
           Salt: 2014.7.2
         Python: 2.7.6 (default, Mar 22 2014, 22:59:56)
         Jinja2: 2.7.3
       M2Crypto: 0.21.1
 msgpack-python: 0.4.4
   msgpack-pure: Not Installed
       pycrypto: 2.6.1
        libnacl: Not Installed
         PyYAML: 3.11
          ioflo: Not Installed
          PyZMQ: 14.0.1
           RAET: Not Installed
            ZMQ: 4.0.4
           Mako: 0.9.1
```
@hundt or @notpeter Are either of you still seeing this issue in the newest release?

**Comment by notpeter** [comment]
Honestly, I gave up on using salt's npm states. 

Ultimately I needed three things it's missing:
 * non global npm install (omit "-g", specify a cwd where node_modules will go)
 * pkgs/versions specified in package.json not in the salt state.
 * install dependencies while not devdependencies ("--production")

All were very easy to handle with a cmd.run so I opted that route for now. Someone could certainly file an ER about those issue, as is npm.installed is pretty weak.

Glad to hear the regression was fixed in 2014.7.2 though.

**Comment by rallytime** [comment]
@notpeter The npm modules and state could certainly use some more functionality. It looks, according to the docs, that there were some new additions and improvements, along with this fix. That being said, I think there is still some work that needs to be done for you to use it as you've specified. Glad you found some work-arounds.

I'm going to close this issue for now, but if this pops up again, please comment and we can open this again. Thanks!

**Comment by rallytime** [comment]
I was finally able to reproduce the error, which I have now fixed in #29715.
